- 在Go编程中，协程是创建计算的唯一途径。

- 协程有时也被称为绿色线程。绿色线程是由程序的运行时（runtime）维护的线程。一个绿色线程的开销（比如内存消耗和情景转换）比一个系统线程常常小得多。 只要内存充足，一个程序可以轻松支持上万个并发协程。

- go语言是不支持创建系统线程的，因此创建协程是Go程序内部唯一的并发实现方式

- 每个Go程序启动的时候只有一个对用户可见的协程，我们称之为主协程。 一个协程可以开启更多其它新的协程。在Go中，开启一个新的协程是非常简单的。 我们只需在一个函数调用之前使用一个`go`关键字，即可让此函数调用运行在一个新的协程之中。 当此函数调用退出后，这个新的协程也随之结束了。我们可以称此函数调用为一个协程调用（或者为此协程的启动调用）。 一个协程调用的所有返回值（如果存在的话）必须被全被舍弃。

- **当主协程退出的时候，其他的程序也会退出，即使还在运行**

- `sync.WaitGroup`和`Done()`、`wait()`、`Add()`的应用

  ```go
  var wg sync.waitGroup
  
  wg.Done() //往往在一个goroutine方法执行完后告诉程序运行完毕，这是为了保证并发的程序执行完而准备的
  wg.Wait() //方法调用将阻塞（等待）到所有任务都已经完成之后才继续执行其后的语句
  wg.Add() //方法用来注册新的需要完成的任务数。
  ```

- 协程的状态

  ![协程状态](https://gfw.go101.org/article/res/goroutine-states.png)

  协程有2种状态，分别是:**运行状态**和**阻塞状态**

  一个协程可以在这两个状态之间切换。 比如上例中的主协程在调用`wg.Wait`方法的时候，将从运行状态切换到阻塞状态；当两个新协程完成各自的任务后，主协程将从阻塞状态切换回运行状态。

  **但需要注意的是一个处于睡眠状态(`time.Sleep`)或者是等待系统调用返回的协程被认为是运行状态，而不是阻塞状态**

  同时，**一种处于阻塞状态的协程不会自发结束阻塞状态**，它必须被另外一个协程通过某种并发同步的方法来被动地结束阻塞状态

- 并非所有处于运行状态的协程都在执行。在任一时刻，只能最多有和逻辑CPU数目一样多的协程在同时执行。 

  可以通过调用`runtime.NumCPU`函数来查询当前程序可以调用的逻辑cpu的数量，下图展示了一个协程的更详细的生命周期

  ![协程详细状态](https://gfw.go101.org/article/res/goroutine-schedule.png)

  在上图中，睡眠或者等待系统调用的状态为运行状态，而不是阻塞状态
  
- M-P-G模型
  
  M:系统线程
  
  P:逻辑处理器(并非逻辑CPU)，来处理大多数的调度工作
  
  G:协程
  
  逻辑处理器像一个监工一样通过将不同的处于运行状态协程（**G**）交给不同的系统线程（**M**）来执行。一个协程在同一时刻只能在一个系统线程中执行。一个执行中的协程运行片刻后将自发地脱离让出一个系统线程，从而使得其它处于等待子状态的协程得到执行机会。

- 可以通过调用`runtime.GOMAXPROCS`来获取和设置逻辑处理器的数量

- 延迟函数调用(deferred function call)

  可以在`defer`关键字后面形成一个延迟函数调用，和协程调用类似，被延迟的函数调用的所有返回值必须全部被舍弃。

  当一个函数调用被延迟后，它不会立即被执行。它将被推入由**当前协程维护(注意是当前的协程)的一个延迟调用堆栈**。 当一个函数调用（可能是也可能不是一个延迟调用）返回并进入它的[退出阶段](https://gfw.go101.org/article/function-declarations-and-calls.html#exiting-phase)后，所有在此函数调用中已经被推入的延迟调用将被按照它们被推入堆栈的顺序逆序执行。 当所有这些延迟调用执行完毕后，此函数调用也就真正退出了。

  如下代码所示:

  ```go
  package main
  
  import "fmt"
  
  func main() {
  	defer fmt.Println("The third line.")
  	defer fmt.Println("The second line.")
  	fmt.Println("The first line.")
  }
  
  /*
  The first line.
  The second line.
  The third line.
  */
  ```

- 每个协程都在维护两个调用堆栈

  1. 一个是正常的函数调用堆栈。在此堆栈中，相邻的两个调用存在着调用关系。晚进入堆栈的调用被早进入堆栈的调用所调用。 此堆栈中最早被推入的调用是对应协程的启动调用。
  2. 另一个堆栈是上面提到的延迟调用堆栈。处于延迟调用堆栈中的任意两个调用之间不存在调用关系。

  ```go
  func main() {
     defer fmt.Println("9")
     fmt.Println("0")
     defer fmt.Println("8")
     fmt.Println("1")
     if false {
        defer fmt.Println("not reachable")
     }
     defer func() {
        defer fmt.Println("7")
        fmt.Println("3")
        defer func() {
           fmt.Println("5")
           fmt.Println("6")
        }()
        fmt.Println("4")
     }()
     fmt.Println("2")
     return
     defer fmt.Println("not reachable")
  }
  ```

  上述代码会顺序输出0-9

- 延迟函数调用可以修改最内层函数的返回值

  下面代码可以修改`Triple()`函数的返回值

  ```go
  func main() {
     fmt.Println(Triple(5))
  }
  
  func Triple(n int) (r int) {
     defer func() {
        r += n
     }()
     return n * 2
  }
  ```

- 一个协程调用或者延迟调用的实参是在此调用发生时被估值的

  - 对于一个延迟函数调用，它的实参是在此调用被推入延迟调用堆栈的时候被估值的。
  - 对于一个协程调用，它的实参是在此协程被创建的时候估值的。

  与上述不同的是，匿名函数体内的表达式实在**该函数被执行时**才会被逐个估值的，**不管此函数是被普通调用还是延迟/协程调用**

  ```go
  func main() {
  	func() {
  		for i := 0; i < 3; i++ {
  			defer fmt.Println("a:", i)
  		}
  	}()
  	fmt.Println()
  	func() {
  		for i := 0; i < 3; i++ {
  			defer func() {
  				fmt.Println("b:", i)
  			}()
  		}
  	}()
  }
  /*
  a: 2
  a: 1
  a: 0
  
  b: 3
  b: 3
  b: 3
  */
  ```

  可以对第二个循环略加修改（使用两种方法），使得它和第一个循环打印出相同的结果。

  ```go
  for i := 0; i < 3; i++ {
  			defer func(i int) {
  				// 此i为形参i，非实参循环变量i。
  				fmt.Println("b:", i)
  			}(i)
  		}
  ```

  或者

  ```go
  for i := 0; i < 3; i++ {
  			i := i // 在下面的调用中，左i遮挡了右i。
  			       // <=> var i = i
  			defer func() {
  				// 此i为上面的左i，非循环变量i。
  				fmt.Println("b:", i)
  			}()
  		}
  ```

  同样的规则也适用于协程调用

  ```go
  package main
  
  import "fmt"
  import "time"
  
  func main() {
  	var a = 123
  	go func(x int) {
  		time.Sleep(time.Second)
  		fmt.Println(x, a) // 123 789
  	}(a)
  
  	a = 789
  
  	time.Sleep(2 * time.Second)
  }
  ```

- go语言不支持异常抛出和捕获，而是推荐使用**返回值显示返回错误**，在go语言中也有类似的异常抛出/捕获机制，被称为恐慌/恢复(panic/recover)机制

  可以调用内置函数`panic`来产生一个恐慌以使当前协程进入恐慌状况。一个恐慌不会蔓延到其它协程。

  函数声明

  ```go
  func panic(v interface{})
  func recover() interface{}
  ```

  

  

  

  

  